service/qos/service_level_controller: Prevent accessing stopped `auth::service`

dawmd opened the pull request:
`auth::service` and its contents are accessed by
`service::qos::service_level_controller` when updating the service level cache
(cf. `service_level_controller::update_effective_service_levels_cache`).
However, `auth::service` is started after the node has jointed the cluster;
see the commit scylladb/scylladb@562caaf6c61f75fb6e1849c70e18fcb57b6aeb46
for more details on that.

Because we keep initialization and de-initialization paired in `main.cc`, that
implies that `auth::service` will be stopped BEFORE `service_level_controller`
will be. As a result, the cache may still be reloaded in between those two events
and effectively lead to a segmentation fault or other issues. We've already
experienced them; see issue: scylladb/scylladb#24792.

To prevent that issue (and potentially other similar ones), we're extending
`auth::service` by a mutex whose sole purpose is to be used by
`service_level_controller` whenever it wants to access it. It's an ugly
solution, but it's much safer than reordering initialization of services
in `main.cc`: we've had enough issues related to that and there's no point
in generating new ones (cf. scylladb/scylladb#2737).

Correctness of the solution:

* The cache of `service_level_controller` cannot be circumvented and as
  a result, we might end up accessing an old state of it. However, it's
  of little consequence since the only scenario affected by these changes
  is when the node is shutting down: some tasks will be run with a lower
  or higher priority than intended, but that's acceptable.
* We may reject operations like finding or dropping a service level, or
  describing the set of them. In those cases, we throw an exception and
  the caller should handle it. Again, since the node is shutting down,
  it should be an acceptable solution.
* The last change this commit brings is that we add validation logic:
  Scylla will crash if we attempt to access `auth::service` when it
  hasn't started yet. Code MUST be sure it can access it. Otherwise,
  we'll only end up with yet another issue only detected by a sanitizer.

Fixes scylladb/scylladb#24792

Backport: We should backport the changes to all affected versions
to prevent invalid memory accesses.

dawmd commented on service/qos/service_level_controller.cc:
No test yet, but should be ready to review.

Jadw1 commented on service/qos/service_level_controller.cc:
```diff
@@ -1042,4 +1057,10 @@ get_service_level_distributed_data_accessor_for_current_version(
+future<std::optional<service_level_controller::auth_service_getter>>
+service_level_controller::get_auth_service() {
+    using unit_type = seastar::semaphore_units<semaphore_default_exception_factory>;
+    co_return auth_service_getter {_auth_service.local(), unit_type{}, };
```
I'd suggest to change order of this and next commit to firstly introduce the semaphore in auth, and then to use it here. Reading only this commit it's unclear why you're introducing the `auth_service_getter` and just using default semaphore units.

Jadw1 commented on auth/service.hh:
```diff
@@ -104,6 +113,30 @@ class service final : public seastar::peering_sharded_service<service> {
+    /// `status::working`. Otherwise, it should avoid doing anything to it.
+    ///
+    /// Ditch this once it stops being necessary.
+    seastar::semaphore _sl_mutex;
```
I think more general name like `_operation_mutex` (or other name that indicates that the service is busy) would be better. Auth doesn't depend on service levels, service levels just use the auth service and imo naming something after SL isn't right.
Theoretically, other services might be using the auth service and may want to keep it alive during their usage.

Jadw1 replied:
> Theoretically, other services might be using the auth service and may want to keep it alive during their usage.

Hmm but then there'll be a queue to use auth which isn't necessary at all :/

dawmd commented:
Version 2:
* Switched from a semaphore to a gate because it feels more natural. Renamed it as suggested in Michał's review.
* Extended comments in the code.
* Ignored the case when we try to access an uninitialised `auth::service` instance during start-up. It will probably be handled as a separate issue. See my messages above for more context.
* Reordered the commits as suggested in Michał's review. Going over the changes should be clearer now.
* Turned into a draft: still no test. I'll add it in the next version.

dawmd commented:
> > The former boils down to not crashing a node when auth::service hasn't been started yet, but service_level_controller demands access to it. We can add a loop with sleeps and be done with it since it only happens at start-up.
> 
> Currently there is a check in `service_level_controller::update_effective_service_levels_cache()` which does early return in case `auth_service` is not initialize yet. I think the loop isn't necessary because the SL cache initialization is done in main.cc, so we won't loose any data.

`auth::service` will be initialised = constructed, but we don't have a guarantee that it's been started yet. I assumed accessing it before that is unsafe, e.g. because we're getting information from the role manager, authenticator, and authoriser, and that information may be inconsistent if some of those services are already running, while some aren't yet. Maybe it's not a problem, though. As far as I can tell, some roles will just not have the right service levels attached to them, but that should be eventually resolved when we reload the cache again (and that *will* happen—we call it explicitly—when the auth service has started). Let's discuss it in a separate issue (I'll create it).

> Besides, the group0 apply fiber shouldn't wait until auth is initialized!

That's true. Even if we somehow avoid updating the cache when reloading the state of the topology, it's still a subtle thing and we cannot allow for prolonging a bootstrap or, even worse, a deadlock...

dawmd commented:
Version 3:
* Added a reproducer.

@xemul please take a look. I'm not sure if the test is worth the bloated code we get in Scylla. Maybe it's better to drop it. If not, I'll update the commit message.

Jadw1 commented on main.cc:
```diff
@@ -2272,7 +2272,72 @@ sharded<locator::shared_token_metadata> token_metadata;
+            //
+            // The ONLY reason why we added step 1 in the actual plan is that it's more convenient.
+            // https://github.com/scylladb/scylladb/issues/24792#issuecomment-3146021819
+            std::any maybe_update_effective_service_levels_cache;
```
I think it looks better if this big comment would be in the test file, and here you can just point to the test

dawmd commented:
> #### 🔴 CI State: FAILURE
> Mode 	Stage 	Status 	Node
> 	[Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18722/) 	✅ 	thor
> dev 	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18722/) 	✅ 	spider1.cloudius-systems.com
> 	[test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18722/) 	❌ 	spider1.cloudius-systems.com
> release 	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18722/) 	❌ 	spider8.cloudius-systems.com
> debug 	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18722/) 	❌ 	spider8.cloudius-systems.com
> #### Failed Tests (3/22290):
> 
>     * [test_try_to_access_stopped_auth_service_when_reloading_effective_sl_cache.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18722/testReport/junit/cluster.auth_cluster/test_raft_service_levels/test_try_to_access_stopped_auth_service_when_reloading_effective_sl_cache_dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_try_to_access_stopped_auth_service_when_reloading_effective_sl_cache.dev.1)
> 
>     * [test_try_to_access_stopped_auth_service_when_reloading_effective_sl_cache.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18722/testReport/junit/cluster.auth_cluster/test_raft_service_levels/test_try_to_access_stopped_auth_service_when_reloading_effective_sl_cache_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_try_to_access_stopped_auth_service_when_reloading_effective_sl_cache.dev.2)
> 
>     * [test_try_to_access_stopped_auth_service_when_reloading_effective_sl_cache.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18722/testReport/junit/cluster.auth_cluster/test_raft_service_levels/test_try_to_access_stopped_auth_service_when_reloading_effective_sl_cache_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_try_to_access_stopped_auth_service_when_reloading_effective_sl_cache.dev.3)
> 
> 
> #### Build Details:
> 
>     * Duration: 4 hr 38 min

It seems that the failed tests are due to `auth::service::stop` being stuck. It's not due to an error injection per se, but from the reproducer being wrong. What probably happened is that `service_level_controller:update_effective_service_levels_cache` got a hold on the access gate of `auth::service` and effectively prevented `auth::service::stop` from closing it. We have a deadlock.

Unfortunately, I don't know if that can be amended. The very premise of the test relies on that... I suggest dropping the test and merging the fix without it.

Jadw1 commented on service/qos/service_level_controller.cc:
```diff
@@ -324,6 +325,13 @@ future<> service_level_controller::update_effective_service_levels_cache() {
+    // Steps 3. and 4. of the plan described in the issue and in `main.cc`.
+    // https://github.com/scylladb/scylladb/issues/24792#issuecomment-3146021819
+    utils::get_local_injector().receive_message("suspend_auth_service_stop");
+    co_await utils::get_local_injector().inject("suspend_update_effective_service_levels_cache_accessing_auth_service",
```
The test fails because updating SL cache hangs here. It's supposed to receive message from `auth/service::stop()` but it never gets to send the message because SL controller holds the auth access gate.

Wouldn't it work, if you do (in the test itself):
- trigger shutdown on a separate thread
- wait for `suspend_auth_service_stop: message received` (or some other log that means that auth is waiting to close the access gate)
- message `suspend_update_effective_service_levels_cache_accessing_auth_service`

If I'm correct, SL controller should finish the update, release the gate and then the auth should stop.
At the same time, the reproducer should reproduce the original error without your fix.

dawmd replied:
If the SL controller finishes the update before `auth::service` stops, the problem will never appear. But you have a good point and I have an idea thanks to it: we could reorder getting access to `auth::service` and the error injection. Then `auth::stop` would finish and only then would we try to access it. It has two pros and one con.

The con is that that way we'll never trigger the same execution order where the problem manifests as long as we have the fix.
The pros are that the test passes on master and that the problem will be triggered if we don't access `auth::service` via the getter I introduced.

Do you think that's an acceptable approach? The code will still be bloated (because iirc we cannot really use the API to enable/message the error injections from the test), but maybe the value outweighs the cost.

Jadw1 commented:
> The problem described in [562caaf](https://github.com/scylladb/scylladb/commit/562caaf6c61f75fb6e1849c70e18fcb57b6aeb46) should be fixed within auth, so it doesn't cause rippling effects in other services.

We discussed the problem with @dawmd and we see 2 ways to fix this:
1. Move initialization of `service_level_controller` after auth is started. This feels right because SL controller depends on auth and not the other way around. However, we were not sure how other services which depends on `service_level_controller`, like `database::start()` would have to be moved a lot later in the main.
2. Extract effective service levels cache to separate entity and remove dependency between `service_level_controller` and auth. It would be a logical step because them SL controller would be responsible for basic service levels (= the data that is in the system table) and the separate cache would depend on the SL controller and auth and it would be calculating effective service levels.

avikivity commented:
> > The problem described in [562caaf](https://github.com/scylladb/scylladb/commit/562caaf6c61f75fb6e1849c70e18fcb57b6aeb46) should be fixed within auth, so it doesn't cause rippling effects in other services.
> 
> We discussed the problem with @dawmd and we see 2 ways to fix this:
> 
> 1. Move initialization of `service_level_controller` after auth is started. This feels right because SL controller depends on auth and not the other way around. However, we were not sure how other services which depends on `service_level_controller`, like `database::start()` would have to be moved a lot later in the main.
> 2. Extract effective service levels cache to separate entity and remove dependency between `service_level_controller` and auth. It would be a logical step because them SL controller would be responsible for basic service levels (= the data that is in the system table) and the separate cache would depend on the SL controller and auth and it would be calculating effective service levels.

In general, factoring a service (in the mathematical sense - factor a number to its prime factors) to get components that break circular dependencies is a direction I like. I'm okay with both choices here as long as the result is clean.
